app/controllers/sessions_controller.rb

```diff
@@ -15,4 +22,34 @@ def set_user_locale
   def default_locale?
     cookies[:locale].eql?(I18n.default_locale)
   end
+
+  def check_recaptcha
+    if recaptcha_required_for_user? && !verify_recaptcha
+      self.resource = resource_class.new sign_in_params
+      redirect_to new_session_path(resource_name, recaptcha: true)
+    else
+      cookies.delete(:login_recaptcha)
+    end
+  end
+
+  def set_recaptcha
+    return unless show_recaptcha?
+
+    @display_recaptcha = true
+    cookies.permanent[:login_recaptcha] = true
+  end
+
+  def show_recaptcha?
+    RecaptchaService.recaptcha_configured? &&
+      (cookies[:login_recaptcha] ||
+      params[:recaptcha] ||
+      recaptcha_required_for_user?)
```

Comment:
This handles three different scenarios...
1. User enters incorrect password, in which case the the email is still in the request parameters, so we can use `recaptcha_required_for_user?` to look up the number of failed attempts
2. If a user tries to login without filling out the recaptcha (which shouldn't really happen since it's disabled), then we redirect to the `new` action with the `recaptcha` param
3. If a user needs a recaptcha, they'll have the `login_recaptcha` cookie set, so if they leave before successfully signing in, then when they return, we'll know to sure the recaptcha

Admittedly, there is probably a cleaner way to handle these scenarios, so I'm open to any suggestions!

Reply:
I don't think I got my environment setup to play around with Recaptcha on local, but from what I'm guessing, would this be the implemented scenario?

**Scenario 1: Incorrect email**

3 failed attempt, we get a cookie to enable a Recaptcha for that session on that specific computer/IP?

**Scenario 2: Correct email, incorrect password**

3 failed attempt, we attach a required Recaptcha to that specified account so no matter what computer someone uses, that account will now require Recaptcha

Case #3 would be an example where someone uses a different browser or refresh the website to not require Recaptcha again, right?

-----

I like that approach, we'll just need to add the tests for each use case. 

At first I was thinking probably just lock it after 3 for that IP, but I suppose we can handle by user basis as well since that is pretty consistent with the current Devise lock settings per user. 

I suppose if the or statements get too big, I suppose you could wrap it in `recaptcha_required?` private method or something to be more intuitive to read?

Reply:
Sorry, I don't think I explained that clearly. All failed attempts are based on username. Right now, there's no tracking of failed attempts on an IP basis. My thought was that a bot could change their IP to get around it, but I can try to add it in if you think it's worthwhile?

The scenarios are really all the same (user tries to log in with the same email with an incorrect password 3 times). I was just trying to explain why each condition in the `if` block was necessary. `params[:recaptcha]` handles the scenarios in which the user redirects to the login page from `redirect_to new_session_path(resource_name, recaptcha: true)` in the `check_recaptcha` up above.  `recaptcha_required_for_user?` handles when a user makes it past that check, but is redirected to the login page by Devise due to an invalid password. Then the cookie is for when the user leaves and comes back. If they cleared the cookie, then they'll get caught by `check_recaptcha` and the form will reload with the recaptcha. Let me know if that's not clear!

Reply:
Ah ok, that makes sense. Good handling on the edge cases 👍

Reply:
I think handling IP would cause scope creep for this PR. I rather stick to what we already have since it's large enough. I think if we think it's worthwhile, we can make another issue if it sounds nice.